Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+            
+        case 10:
+            // ALTSVC frame : RFC 7838 § 4
+            guard header.length >= 2 else {
+                // Must be at least two bytes, to contain the length of the optional 'Origin' field.
+                throw NIOHTTP2Errors.ConnectionError(code: .frameSizeError)
+            }
+            
+            let originLen: UInt16 = bytes.readInteger()!
+            let origin: String?
+            if originLen > 0 {
+                origin = bytes.readString(length: Int(originLen))!
+            } else {
+                origin = nil
+            }
+            
+            let fieldLen = header.length - 2 - Int(originLen)
+            let value: ByteBuffer?
+            if fieldLen != 0 {
+                value = bytes.readSlice(length: fieldLen)!
+            } else {
+                value = nil
+            }
+            
+            payload = .alternativeService(origin: origin, field: value)
+            
+        case 11:
+            // ORIGIN frame : RFC 8336 § 2
+            guard streamID == .rootStream else {
+                // The ORIGIN frame MUST be sent on stream 0; an ORIGIN frame on any
+                // other stream is invalid and MUST be ignored.
+                return nil
+            }
+            
+            var origins: [String] = []
+            var remaining = header.length
+            while remaining > 0 {
+                guard remaining >= 2 else {
+                    // If less than two bytes remain, this is a malformed frame.
+                    throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+                }
+                let originLen: UInt16 = bytes.readInteger()!
+                remaining -= 2
+                
+                guard remaining >= Int(originLen) else {
+                    // Malformed frame.
+                    throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+                }
+                let origin = bytes.readString(length: Int(originLen))!
+                remaining -= Int(originLen)
+                
+                origins.append(origin)
+            }
+            
+            payload = .origin(origins)
+            
+        default:
+            // RFC 7540 § 4.1 https://httpwg.org/specs/rfc7540.html#FrameHeader
+            //    "Implementations MUST ignore and discard any frame that has a type that is unknown."
+            self.state = .idle
+            bytes.moveReaderIndex(forwardBy: header.length)
+            return nil
+        }
+        
+        self.state = .idle
+        let frame = HTTP2Frame(streamID: streamID, flags: flags, payload: payload)
+        
+        switch payload {
+        case .headers where !flags.contains(.endHeaders), .pushPromise where !flags.contains(.endHeaders):
+            // don't emit the frame, store it and wait for CONTINUATION frames to arrive
+            self.currentHeaderFrame = frame
+            return nil
```

Comment:
Also, all the other cases should be checking that this value is `nil`, and throwing if it isn't. That will let us detect violations of RFC 7540.